### ISSUE TITLE
xgbTree with classProbs = TRUE is completely broken on current (Dec 7) github master branch

### ISSUE THREAD

**Issue by hadjipantelis** [topepo/caret#549]
Hello,

This is directly related to [511](https://github.com/topepo/caret/issues/511) and the submit [d992fad](https://github.com/JackStat/caret/commit/d992fadc69dc7fdab3342cb1ab350ab95105ebdf) and the related commit [ed79d6d](https://github.com/topepo/caret/commit/ed79d6d41fe8cef0682a58527a6e3f85c5d3fbc0); it summarily kills `xgboost` when setting ` classProbs = TRUE`. All submits after that are broken for method `xgbTree`.

``` R
remove.packages('caret')
devtools::install_github('topepo/caret/pkg/caret')
library(caret)
 
set.seed(42)  
trainData <- twoClassSim(500)    
testData <- twoClassSim(500)  

trCntlList <- trainControl(method = "boot", number = 6, classProbs = TRUE) 

set.seed(55)  
xgbFitError1 <- train(  Class ~ ., data = trainData,  objective = "binary:logistic", 
                   method = "xgbTree", tuneLength = 2, trControl =  trCntlList)
set.seed(55)  
xgbFitError2 <- train( Class ~ ., data = trainData, objective = "binary:logitraw",
                   method = "xgbTree", tuneLength = 2, trControl =  trCntlList) 

trCntlList <- trainControl(method = "boot", number = 6, classProbs = FALSE) 

set.seed(55)  
xgbFitOK1<- train(  Class ~ ., data = trainData,  objective = "binary:logistic", 
                   method = "xgbTree", tuneLength = 2, trControl =  trCntlList)
set.seed(55)  
xgbFitOK2 <- train( Class ~ ., data = trainData, objective = "binary:logitraw",
                   method = "xgbTree", tuneLength = 2, trControl =  trCntlList) 
```

The error occurs within the `foreach` loop of `nominalTrainWorkflow`. In particular around Line # ~203 during the call to the `probFunction`. `probFunction` errs because the argument `modelFit$param$objective` is NULL. To that respect, `modelFit` does not have a `param` field to begin with.  That is somewhat surprising but that it can be explained by the fact that `createModel` never appended a `param` field. This is expected as `createModel` is fully agnostic about the objective function used. It should have a point where it creates a `params` field to the `modelFit` returned.

I can tell that this error was introduced with `ed79d6d` as I back-tracked it; the code runs "fine" at the `5cc19a5` commit. (well, not totally fine, the probabilities are wrong as show in  [511](https://github.com/topepo/caret/issues/511)  but at least it works when `objective = "binary:logistic"`). My `sessionInfo()` is as follows:
  
``` R
R version 3.3.1 (2016-06-21)
Platform: x86_64-redhat-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C               LC_TIME=en_GB.UTF-8       
 [4] LC_COLLATE=en_GB.UTF-8     LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
 [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] caret_6.0-73    plyr_1.8.4      xgboost_0.4-4   ggplot2_2.2.0   lattice_0.20-33

loaded via a namespace (and not attached):
 [1] reshape2_1.4.2     splines_3.3.1      colorspace_1.2-6   testthat_1.0.2    
 [5] stats4_3.3.1       mgcv_1.8-12        chron_2.3-47       ModelMetrics_1.1.0
 [9] e1071_1.6-7        nloptr_1.0.4       withr_1.0.2        foreach_1.4.3     
[13] stringr_1.1.0      MatrixModels_0.4-1 munsell_0.4.3      gtable_0.2.0      
[17] devtools_1.12.0    codetools_0.2-14   memoise_1.0.0      SparseM_1.72      
[21] quantreg_5.29      pbkrtest_0.4-6     parallel_3.3.1     curl_1.2          
[25] class_7.3-14       Rcpp_0.12.8        scales_0.4.1       lme4_1.1-12       
[29] digest_0.6.10      stringi_1.1.1      grid_3.3.1         tools_3.3.1       
[33] magrittr_1.5       lazyeval_0.2.0     tibble_1.2         crayon_1.3.2      
[37] car_2.1-4          MASS_7.3-45        Matrix_1.2-6       data.table_1.9.6  
[41] assertthat_0.1     minqa_1.2.4        roxygen2_5.0.1     httr_1.2.1        
[45] iterators_1.0.8    R6_2.2.0           nnet_7.3-12        nlme_3.1-128      
[49] compiler_3.3.1     git2r_0.15.0       
```
Immediate work-around: I believe that adding the line `modelFit$params <- list(...)` within `createModel` will be adequate. Otherwise one would have to pull the objective function used out of the "xgb.Booster" output of `xgb.train` and that seems to be complicated. 

I will create a relevant pull request.

All the best,
Pantelis

**Comment by topepo** [comment]
Sit tight; I believe that this is an issue with some changes related to getting the optimism bootstrap estimate (issue #546).

**Comment by hadjipantelis** [comment]
Cool, best of luck!

As you saw I pushed a fix earlier today to fix this issue on master - pull request #551. `multiSummary` was totally unusable at the time. (I did not mention that in the body as even simpler cases with `twoClassSummary` crushed.)

**Comment by topepo** [comment]
This all works now, so I'll close the issue
